gh-100690: Prevent prefix "called_" for attributes of safe mock objects

cklein opened the pull request:
Prevent prefix "called_" for mocks in safe mode

- Issue: gh-100690

cjw296 reviewed (commented):
I'm -0 on this change, feels like we're into the diminishing realms where someone out there is going to have a legitimate use case for `called_name` or some such and this then prevents them using a safe mock.

Who else is hitting this?

cklein committed:
Prevent calling assert functions without the prefix "assert_" in safe mode

Raise an AttributeError when calling e.g. `mock_obj.called_once` instead of
`mock_obj.assert_called_once`.

cklein committed:
Merge remote-tracking branch 'python/main' into disallow_mock_method_prefix

cklein committed:
Added tests to show that names from denylist still can be used with a spec

cjw296 commented on Lib/unittest/mock.py:
```diff
@@ -1062,6 +1062,10 @@ def _calls_repr(self, prefix="Calls"):
         return f"\n{prefix}: {safe_repr(self.mock_calls)}."
 
 
+# gh-100690 Denylist for forbidden attribute names in safe mode
```
Personally I dislike these kind of bug links, the comment is fine without it, and in a decade's time when we're no longer using GitHub for issue tracking, this will be a wart ;-)

merwok replied:
Same for me

cjw296 commented on Lib/unittest/mock.py:
```diff
@@ -1062,6 +1062,10 @@ def _calls_repr(self, prefix="Calls"):
 
 
+# gh-100690 Denylist for forbidden attribute names in safe mode
+ATTRIB_DENY_LIST = {name.removeprefix("assert_") for name in dir(NonCallableMock) if name.startswith("assert_")}
```
Much prefer this approach! 🎉

cklein committed:
Changed documentation

merwok commented on Misc/NEWS.d/next/Library/2023-01-02-16-59-49.gh-issue-100690.2EgWPS.rst:
```diff
@@ -0,0 +1,3 @@
+Mock objects which are not unsafe will now raise an AttributeError when accessing an
+attribute that matches the name of an assertion but without the prefix ``assert_``, e.g. accessing ``called_once`` instead of ``assert_called_once``.
+This is in addition to this already happening for accessing attributes with prefixes assert, assret, asert, aseert, and assrt.
```
Minor thing: can you wrap to 72 columns?  And add markup to Mock and AttributeError, plus code markup for the prefixes on last line.

cklein committed:
Reformatted NEWS

cjw296 committed:
Merge branch 'main' into disallow_mock_method_prefix

sobolevn commented on Lib/unittest/mock.py:
```diff
@@ -1062,6 +1062,10 @@ def _calls_repr(self, prefix="Calls"):
 
 
+# Denylist for forbidden attribute names in safe mode
+ATTRIB_DENY_LIST = {name.removeprefix("assert_") for name in dir(NonCallableMock) if name.startswith("assert_")}
```
Sorry for being late to the party, but anyways :)

I think that `ATTRIB_DENY_LIST` is an implementation detail, so it should be `_ATTRIB_DENY_LIST`. And also, it is not suited for mutation, isn't it? So, it should be a `frozenset` instead :)

sobolevn replied:
Since this PR is already merged, I've opened https://github.com/python/cpython/pull/100819 with this proposal. Feel free to close if it should be public and mutable!

merwok replied:
It is already non-public by not being included in `__all__`, and the mutability is not really an issue.
